# Olist Data Cleaning

### Importing Libraries

In [3]:
import os
import pandas as pd
import numpy as np

### creating data frames from olist csv files

In [4]:
# Importing Datasets from download dir
dataset_path = r"olist_dataset"
csv_files = os.listdir(dataset_path)
olistbr_datasets = {}
for i in csv_files:
    olistbr_datasets[i.replace('.csv', '')] = pd.read_csv(fr"{dataset_path}\{i}")
dataset_keys = list(olistbr_datasets.keys())
print(dataset_keys)

['olist_customers_dataset', 'olist_geolocation_dataset', 'olist_orders_dataset', 'olist_order_items_dataset', 'olist_order_payments_dataset', 'olist_order_reviews_dataset', 'olist_products_dataset', 'olist_sellers_dataset', 'product_category_name_translation']


In [5]:
# Reading all donwloaded csv files and converting into Pandas Data Frame
olist_customers_dataset = pd.read_csv(fr"{dataset_path}\{dataset_keys[0]}.csv")
olist_geolocation_dataset = pd.read_csv(fr"{dataset_path}\{dataset_keys[1]}.csv")
olist_order_dataset = pd.read_csv(fr"{dataset_path}\{dataset_keys[2]}.csv")
olist_order_items_dataset = pd.read_csv(fr"{dataset_path}\{dataset_keys[3]}.csv")
olist_order_payments_dataset = pd.read_csv(fr"{dataset_path}\{dataset_keys[4]}.csv")
olist_order_reviews_dataset = pd.read_csv(fr"{dataset_path}\{dataset_keys[5]}.csv")
olist_products_dataset = pd.read_csv(fr"{dataset_path}\{dataset_keys[6]}.csv")
olist_sellers_dataset = pd.read_csv(fr"{dataset_path}\{dataset_keys[7]}.csv")
product_category_name_translation = pd.read_csv(fr"{dataset_path}\{dataset_keys[8]}.csv")

### finds which dataset contains how much null values

In [5]:
df = olist_customers_dataset.copy()

# Define all possible null tokens
null_tokens = ['N/A', 'n/a', 'NA', 'na', 'none', 'null', 'Null', 'None', '']

# Clean and standardize text fields
df = df.map(lambda x: x.strip().lower() if isinstance(x, str) else x)

# Replace text-based nulls with actual np.nan
df.replace(null_tokens, np.nan, inplace=True)

# Summary of missing values per column
missing_summary = df.isna().sum().reset_index()
missing_summary.columns = ['Column', 'Missing_Values']

print(missing_summary)


                     Column  Missing_Values
0               customer_id               0
1        customer_unique_id               0
2  customer_zip_code_prefix               0
3             customer_city               0
4            customer_state               0


In olist_customers_dataset 0 Null values precent

In [6]:
df = olist_geolocation_dataset.copy()

# Define all possible null tokens
null_tokens = ['N/A', 'n/a', 'NA', 'na', 'none', 'null', 'Null', 'None', '']

# Clean and standardize text fields
df = df.map(lambda x: x.strip().lower() if isinstance(x, str) else x)

# Replace text-based nulls with actual np.nan
df.replace(null_tokens, np.nan, inplace=True)

# Summary of missing values per column
missing_summary = df.isna().sum().reset_index()
missing_summary.columns = ['Column', 'Missing_Values']

print(missing_summary)


                        Column  Missing_Values
0  geolocation_zip_code_prefix               0
1              geolocation_lat               0
2              geolocation_lng               0
3             geolocation_city               0
4            geolocation_state               0


In olist_geolocation_dataset 0 Null values precent

In [7]:
df = olist_order_dataset.copy()

# Define all possible null tokens
null_tokens = ['N/A', 'n/a', 'NA', 'na', 'none', 'null', 'Null', 'None', '']

# Clean and standardize text fields
df = df.map(lambda x: x.strip().lower() if isinstance(x, str) else x)

# Replace text-based nulls with actual np.nan
df.replace(null_tokens, np.nan, inplace=True)

# Summary of missing values per column
missing_summary = df.isna().sum().reset_index()
missing_summary.columns = ['Column', 'Missing_Values']

print(missing_summary)


                          Column  Missing_Values
0                       order_id               0
1                    customer_id               0
2                   order_status               0
3       order_purchase_timestamp               0
4              order_approved_at             160
5   order_delivered_carrier_date            1783
6  order_delivered_customer_date            2965
7  order_estimated_delivery_date               0


In [8]:
olist_order_dataset.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In olist_order_dataset had null values
order_approved_at = 160
order_delivered_carrier_date = 1783
order_delivered_customer_date = 2965

In [9]:
df = olist_order_items_dataset.copy()

# Define all possible null tokens
null_tokens = ['N/A', 'n/a', 'NA', 'na', 'none', 'null', 'Null', 'None', '']

# Clean and standardize text fields
df = df.map(lambda x: x.strip().lower() if isinstance(x, str) else x)

# Replace text-based nulls with actual np.nan
df.replace(null_tokens, np.nan, inplace=True)

# Summary of missing values per column
missing_summary = df.isna().sum().reset_index()
missing_summary.columns = ['Column', 'Missing_Values']

print(missing_summary)


                Column  Missing_Values
0             order_id               0
1        order_item_id               0
2           product_id               0
3            seller_id               0
4  shipping_limit_date               0
5                price               0
6        freight_value               0


In olist_order_items_dataset 0 Null values precent

In [10]:
df = olist_order_payments_dataset.copy()

# Define all possible null tokens
null_tokens = ['N/A', 'n/a', 'NA', 'na', 'none', 'null', 'Null', 'None', '']

# Clean and standardize text fields
df = df.map(lambda x: x.strip().lower() if isinstance(x, str) else x)

# Replace text-based nulls with actual np.nan
df.replace(null_tokens, np.nan, inplace=True)

# Summary of missing values per column
missing_summary = df.isna().sum().reset_index()
missing_summary.columns = ['Column', 'Missing_Values']

print(missing_summary)


                 Column  Missing_Values
0              order_id               0
1    payment_sequential               0
2          payment_type               0
3  payment_installments               0
4         payment_value               0


In olist_order_payments_dataset 0 Null values precent

In [11]:
df = olist_order_reviews_dataset.copy()

# Define all possible null tokens
null_tokens = ['N/A', 'n/a', 'NA', 'na', 'none', 'null', 'Null', 'None', '']

# Clean and standardize text fields
df = df.map(lambda x: x.strip().lower() if isinstance(x, str) else x)

# Replace text-based nulls with actual np.nan
df.replace(null_tokens, np.nan, inplace=True)

# Summary of missing values per column
missing_summary = df.isna().sum().reset_index()
missing_summary.columns = ['Column', 'Missing_Values']

print(missing_summary)


                    Column  Missing_Values
0                review_id               0
1                 order_id               0
2             review_score               0
3     review_comment_title           87658
4   review_comment_message           58275
5     review_creation_date               0
6  review_answer_timestamp               0


In olist_order_reviews_dataset had null values : review_comment_title = 87658 and review_comment_message = 58275

In [12]:
df = olist_products_dataset.copy()

# Define all possible null tokens
null_tokens = ['N/A', 'n/a', 'NA', 'na', 'none', 'null', 'Null', 'None', '']

# Clean and standardize text fields
df = df.map(lambda x: x.strip().lower() if isinstance(x, str) else x)

# Replace text-based nulls with actual np.nan
df.replace(null_tokens, np.nan, inplace=True)

# Summary of missing values per column
missing_summary = df.isna().sum().reset_index()
missing_summary.columns = ['Column', 'Missing_Values']

print(missing_summary)


                       Column  Missing_Values
0                  product_id               0
1       product_category_name             610
2         product_name_lenght             610
3  product_description_lenght             610
4          product_photos_qty             610
5            product_weight_g               2
6           product_length_cm               2
7           product_height_cm               2
8            product_width_cm               2


In [13]:
olist_products_dataset.isnull().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

In olist_products_dataset had some null values 610 - 610 null values in 4 columns and 2 - 2 null values in 4 columns

In [14]:
df = olist_sellers_dataset.copy()

# Define all possible null tokens
null_tokens = ['N/A', 'n/a', 'NA', 'na', 'none', 'null', 'Null', 'None', '']

# Clean and standardize text fields
df = df.map(lambda x: x.strip().lower() if isinstance(x, str) else x)

# Replace text-based nulls with actual np.nan
df.replace(null_tokens, np.nan, inplace=True)

# Summary of missing values per column
missing_summary = df.isna().sum().reset_index()
missing_summary.columns = ['Column', 'Missing_Values']

print(missing_summary)


                   Column  Missing_Values
0               seller_id               0
1  seller_zip_code_prefix               0
2             seller_city               0
3            seller_state               0


In olist_sellers_dataset 0 Null values precent

In [15]:
df = product_category_name_translation.copy()

# Define all possible null tokens
null_tokens = ['N/A', 'n/a', 'NA', 'na', 'none', 'null', 'Null', 'None', '']

# Clean and standardize text fields
df = df.map(lambda x: x.strip().lower() if isinstance(x, str) else x)

# Replace text-based nulls with actual np.nan
df.replace(null_tokens, np.nan, inplace=True)

# Summary of missing values per column
missing_summary = df.isna().sum().reset_index()
missing_summary.columns = ['Column', 'Missing_Values']

print(missing_summary)


                          Column  Missing_Values
0          product_category_name               0
1  product_category_name_english               0


In product_category_name_translation 0 Null values precent

### Handling Missing or null values from olist_products_dataset

In [16]:
olist_products_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


In [17]:
# filling null values with unknown in product_category_name column
olist_products_dataset['product_category_name'].fillna('unknown', inplace=True)

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_3976\303500999.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  olist_products_dataset['product_category_name'].fillna('unknown', inplace=True)


In [18]:
# filling null values with 0 in all other columns which contains null values
for column in ['product_name_lenght', 'product_description_lenght','product_photos_qty']:
    olist_products_dataset[column].fillna(0, inplace=True)

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_3976\1674758920.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  olist_products_dataset[column].fillna(0, inplace=True)


In [19]:
# Droping nan values from columns which had only 2 null values precents
olist_products_dataset.dropna(subset=['product_weight_g', 'product_length_cm', 'product_height_cm', 'product_width_cm'], inplace=True)

In [20]:
# Verify the cleaning
print("Missing values in olist_products_dataset after cleaning:")
print(olist_products_dataset.isna().sum())

Missing values in olist_products_dataset after cleaning:
product_id                    0
product_category_name         0
product_name_lenght           0
product_description_lenght    0
product_photos_qty            0
product_weight_g              0
product_length_cm             0
product_height_cm             0
product_width_cm              0
dtype: int64


### Checking and converting Data types

#### Olist Order dataset

In [21]:
olist_order_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB


In [22]:
olist_order_dataset.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

In [23]:
order_data_cols = [
    'order_purchase_timestamp',
    'order_approved_at',
    'order_delivered_carrier_date',
    'order_delivered_customer_date',
    'order_estimated_delivery_date'
]

for col in order_data_cols:
    olist_order_dataset[col] = pd.to_datetime(olist_order_dataset[col])

In [24]:
# checking
olist_order_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       99441 non-null  object        
 1   customer_id                    99441 non-null  object        
 2   order_status                   99441 non-null  object        
 3   order_purchase_timestamp       99441 non-null  datetime64[ns]
 4   order_approved_at              99281 non-null  datetime64[ns]
 5   order_delivered_carrier_date   97658 non-null  datetime64[ns]
 6   order_delivered_customer_date  96476 non-null  datetime64[ns]
 7   order_estimated_delivery_date  99441 non-null  datetime64[ns]
dtypes: datetime64[ns](5), object(3)
memory usage: 6.1+ MB


#### Order Items Dataset

In [25]:
olist_order_items_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


In [26]:
olist_order_items_dataset.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [27]:
olist_order_items_dataset['shipping_limit_date'] = pd.to_datetime(olist_order_items_dataset['shipping_limit_date'])

In [28]:
olist_order_items_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   order_id             112650 non-null  object        
 1   order_item_id        112650 non-null  int64         
 2   product_id           112650 non-null  object        
 3   seller_id            112650 non-null  object        
 4   shipping_limit_date  112650 non-null  datetime64[ns]
 5   price                112650 non-null  float64       
 6   freight_value        112650 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(1), object(3)
memory usage: 6.0+ MB


#### Order Reviews Dataset

In [29]:
olist_order_reviews_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   review_id                99224 non-null  object
 1   order_id                 99224 non-null  object
 2   review_score             99224 non-null  int64 
 3   review_comment_title     11568 non-null  object
 4   review_comment_message   40977 non-null  object
 5   review_creation_date     99224 non-null  object
 6   review_answer_timestamp  99224 non-null  object
dtypes: int64(1), object(6)
memory usage: 5.3+ MB


In [30]:
olist_order_reviews_dataset.head()

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53

In [31]:
order_data_cols = [
    'review_creation_date',
    'review_answer_timestamp'
]

for col in order_data_cols:
    olist_order_reviews_dataset[col] = pd.to_datetime(olist_order_reviews_dataset[col])

In [32]:
olist_order_reviews_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   review_id                99224 non-null  object        
 1   order_id                 99224 non-null  object        
 2   review_score             99224 non-null  int64         
 3   review_comment_title     11568 non-null  object        
 4   review_comment_message   40977 non-null  object        
 5   review_creation_date     99224 non-null  datetime64[ns]
 6   review_answer_timestamp  99224 non-null  datetime64[ns]
dtypes: datetime64[ns](2), int64(1), object(4)
memory usage: 5.3+ MB


### Saving Cleaned Datasets

In [ ]:
# Creating dirts
cleaned_dataset_path = 'olist_dataset_cleaned'
os.makedirs(cleaned_dataset_path, exist_ok=True)

In [54]:
all_dataframes = {
    'olist_customers_dataset.csv': olist_customers_dataset,
    'olist_geolocation_dataset.csv': olist_geolocation_dataset,
    'olist_orders_dataset.csv': olist_order_dataset, # Modified (datetime)
    'olist_order_items_dataset.csv': olist_order_items_dataset, # Modified (datetime)
    'olist_order_payments_dataset.csv': olist_order_payments_dataset,
    'olist_order_reviews_dataset.csv': olist_order_reviews_dataset, # Modified (datetime)
    'olist_products_dataset.csv': olist_products_dataset, # Modified (nulls)
    'olist_sellers_dataset.csv': olist_sellers_dataset,
    'product_category_name_translation.csv': product_category_name_translation
}

In [55]:
for filename, df in all_dataframes.items():
    output_path = os.path.join(cleaned_dataset_path, filename)
    
    df.to_csv(output_path, index=False)
    
    print(f"Successfully saved {filename} to {output_path}")

print("\nAll cleaned data files have been saved successfully!")

Successfully saved olist_customers_dataset.csv to olist_dataset_cleaned\olist_customers_dataset.csv
Successfully saved olist_geolocation_dataset.csv to olist_dataset_cleaned\olist_geolocation_dataset.csv
Successfully saved olist_orders_dataset.csv to olist_dataset_cleaned\olist_orders_dataset.csv
Successfully saved olist_order_items_dataset.csv to olist_dataset_cleaned\olist_order_items_dataset.csv
Successfully saved olist_order_payments_dataset.csv to olist_dataset_cleaned\olist_order_payments_dataset.csv
Successfully saved olist_order_reviews_dataset.csv to olist_dataset_cleaned\olist_order_reviews_dataset.csv
Successfully saved olist_products_dataset.csv to olist_dataset_cleaned\olist_products_dataset.csv
Successfully saved olist_sellers_dataset.csv to olist_dataset_cleaned\olist_sellers_dataset.csv
Successfully saved product_category_name_translation.csv to olist_dataset_cleaned\product_category_name_translation.csv

All cleaned data files have been saved successfully!
